src/mlpack/methods/ann/layer/batchnorm_impl.hpp

```diff
@@ -0,0 +1,136 @@
+/**
+ * @file batchnorm_impl.hpp
+ * @author Praveen Ch
+ *
+ * Implementation of the Batch Normalization Layer.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "batchnorm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template <typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm(
+    const size_t size, const double eps) : size(size), eps(eps)
+{
+  weights.set_size(size + size, 1);
+}
+
+
+template<typename InputDataType, typename OutputDataType>
+void BatchNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+  deterministic = false;
+  gamma.fill(1.0);
+  beta.fill(0.0);
+  stats.reset();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BatchNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  output.reshape(input.n_rows, input.n_cols);
+
+  // Mean and variance over the entire training set will be used to compute
+  // the forward pass when deterministic is set to true.
+  if (deterministic)
+  {
+    mean = stats.mean();
+    variance = stats.var(1);    
+  }
+  else
+  {
+    mean = arma::mean(input, 1);
+    variance = arma::var(input, 1, 1);
+    
+    for (size_t i = 0; i < output.n_cols; i++)
+    {
+      stats(input.col(i));
+    }
+  }
+
+    for (size_t i = 0; i < output.n_rows; i++)
+    {
+      output.row(i) = input.row(i) - arma::as_scalar(mean.row(i));
+      output.row(i) /= (arma::as_scalar(arma::sqrt(variance.row(i) + eps)));
+      output.row(i) *= arma::as_scalar(gamma.row(i));
```

Comment:
Should we use '.each_row' function here? I am not sure how its implemented in Armadillo, but Armadillo is known to implement optimisations everywhere. Maybe it is using some clever row iterator form. Nonetheless, I think it will be at least as good as the implementation we are doing. What do you think?